### PR TITLE
packagekit, systemd: Use notifications for available updates info

### PR DESCRIPTION
**Pull request by mvollmer**
Now the host overview page gets all information about available
software updates from the "Software Updates" page itself, via page
status notifications.

### CODE REVIEW

**Comment by mvollmer** [comment]
There are differences in detail:

 - There is no distinction between "Enhancements" and "Bug fixes".  The software udpates page itself doesn't make that distinction at the top level, so Ileft that out on the "System" page as well.  Should be easy to bring back if we want it.

- No more progress spinner.

- The "System" page now also shows "Restart recommended" when the "Software Updates" page shows that.  We can extend that to other states, such as "Updates are being applied", etc.

**Comment by mvollmer** [comment]
> There are differences in detail:
> 
>     * There is no distinction between "Enhancements" and "Bug fixes".  The software udpates page itself doesn't make that distinction at the top level, so Ileft that out on the "System" page as well.  Should be easy to bring back if we want it.

I have changed this now to use the old texts and distinctions.  For some reason, the "dnf" backend now seems to properly recognize severities, so the tests needed some adaptation anyway.

**Comment by mvollmer** [comment]
As the test failures of rhel-8-0-distropkg piont out, this relies on a recent enough shell so the /updates is actually pre-loaded.  Are we fine with that? @martinpitt?

**Comment by martinpitt** [comment]
@mvollmer: Yes, that's fine. If at all, we'll do selective backports to our 8.0 packages, we won't rebase. We may rebase cockpit-packagekit for 7.8 extras though, but not cockpit-system (but that won't affect the System page, just SW updates). For testing it's just important that the SW Updates page handles the absence of notifications gracefully, i. e. it will just not actually send anything or get preloaded, but there is no crash.

**Comment by martinpitt** [comment]
Unblocking, this now needs to be rebased.

**Comment by mvollmer** [comment]
> Unblocking, this now needs to be rebased.

Done.

**Comment by martinpitt** [comment]
The [8.0-distropkg tests](https://209.132.184.41:8493/logs/pull-12620-20190902-140404-f27c1be0-cockpit-project-cockpit--rhel-8-0-distropkg/log.html) need to be adjusted. Actually at this point we can just drop the 8-0 tests for master and introduce 8-1-distropkg. I'll work on that on the side.

File: pkg/packagekit/updates.jsx [hunk 1]
```diff
@@ -142,6 +143,14 @@ function count_security_updates(updates) {
     return num_security;
 }
 
+function find_highest_severity(updates) {
+    var max = PK.Enum.INFO_LOW;
+    for (const u in updates)
+        if (updates[u].severity > max)
+            max = updates[u].severity;
+    return max;
+}
```
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: This would fit well right next to `getSeverityIcon()` in pkg/lib/packagekit.js instead.

File: pkg/packagekit/updates.jsx [hunk 2]
```diff
@@ -556,13 +565,14 @@ class OsUpdates extends React.Component {
                                 console.warn("GetTransactionList: failed to read PackageKit transaction roles:", ex.message);
                                 // be robust, try to continue with loading updates anyway
                                 this.initialLoadOrRefresh();
+                                this.loadUpdates();
```
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: What dose that do? initialLoadOrRefresh() already calls loadUpdates(). Also, this is only the error handling for the (Should Not Happen™) case that packagekit is very broken. This seems unrelated to this notification work?

File: pkg/packagekit/updates.jsx [hunk 3]
```diff
@@ -556,13 +565,14 @@ class OsUpdates extends React.Component {
...
                             });
                 })
                 .fail(this.handleLoadError);
     }
 
     handleLoadError(ex) {
-        console.error("loading available updates failed:", JSON.stringify(ex));
+        console.warn("loading available updates failed:", JSON.stringify(ex));
```
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: This was actually deliberately an error so that the tests would fail if it updates failed in some case. This is why one test had `allow_browser_errors()`. If you don't like this much, I don't have a strong attachment to it, but it should be a separate commit then.

File: pkg/packagekit/updates.jsx [hunk 4]
```diff
@@ -682,15 +692,29 @@ class OsUpdates extends React.Component {
...
+        else
+            this.loadUpdates();
+    }
+
+    loadOrRefresh(always_load) {
         PK.call("/org/freedesktop/PackageKit", "org.freedesktop.PackageKit", "GetTimeSinceAction",
                 [PK.Enum.ROLE_REFRESH_CACHE])
-                .done(seconds => {
+                .done(results => {
+                    const seconds = results[0];
```
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: If `results` is  a list here, I'm a bit stumped how it worked before? It seems to me that the "refresh indexes if they are older than a day" logic actually worked in practice.

File: pkg/packagekit/updates.jsx [hunk 5]
```diff
@@ -785,6 +818,9 @@ class OsUpdates extends React.Component {
                 </div>);
         }
 
+        if (["available", "uptodate", "updateSuccess"].indexOf(this.state.state) == -1)
+            page_status.set_own(null);
```
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: This looks backwards -- aren't these precisely the "good" states where we *do* want to clear the notification? I. e. shouldn't this be `>= 0`? Or did I misunderstand what `set_own(null)` does? In the latter case this should have a comment.

File: pkg/systemd/index.html [hunk 6]
```diff
@@ -99,7 +99,7 @@
                 <span id="system_information_os_text"></span>
 
                 <div role="group" class="system-information-updates">
-                    <span id="system_information_updates_icon" hidden></span>
+                    <span id="system_information_updates_icon"></span>
```
Comment (martinpitt) [inline pkg/systemd/index.html]: Was that deliberate? To avoid the text jumping around? This should be split out into a separate commit with justification (FTR, I agree, but it should be explicit).

**Comment by martinpitt** [comment]
This now is conflict-y, and it seems to me that this should also use the new `PageStatusNotifications` from #12600, so I mark this blocked.

File: pkg/packagekit/updates.jsx (hunk 2 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: It's a leftover from some earlier attempts.  Good catch, I'll remove it.

File: pkg/packagekit/updates.jsx (hunk 3 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: When this is a error, all (or at leat a lot) tests will fail since we now pre-load the updates page.  Hmm.

File: pkg/packagekit/updates.jsx (hunk 4 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: Yes, it did work, because of JS magic. `[ 123 ] < 124` => `true`.  I guess this should be a separate commit.

File: pkg/packagekit/updates.jsx (hunk 3 above)
Comment (martinpitt) [reply pkg/packagekit/updates.jsx]: Right, makes sense. Thanks!

File: pkg/packagekit/updates.jsx (hunk 4 above)
Comment (martinpitt) [reply pkg/packagekit/updates.jsx]: OMGponies -- thanks for the explanation!

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: These are the states that call `page_status.set_own()` themselves further down.  An alternative would be to insert `set_own(null)` into all cases of the switch.  That might indeed be less confusing.

File: pkg/systemd/index.html (hunk 6 above)
Comment (mvollmer) [reply pkg/systemd/index.html]: It's not necessary, the visibility is immediately determined by the code upon loading.

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: I have done that, and I like it much better.

File: pkg/packagekit/updates.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: The elements of `updates` are custom structs created in this file, so I don't think lib/packagekit.js should know about them.

Comment (martinpitt) [reply pkg/packagekit/updates.jsx]: Ack.

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (martinpitt) [reply pkg/packagekit/updates.jsx]: That avoids the indexOf(), but I'm still wondering the same thing: Why do the "bad" states clear the notification, and the "good" ones don't? It feels like it should be the other way around.

**Comment by martinpitt** [comment]
Ack, I now [see](https://209.132.184.41:8493/logs/pull-12620-20190910-130303-47b93daf-cockpit-project-cockpit--fedora-30/log.html#30-2) what you meant with the console.error(). So please revert this back to `console.warn()`, as you had before.

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (mvollmer) [reply pkg/packagekit/updates.jsx]: A page status isn't always bad.  For example, "uptodate" sets the status to "System is up to date" (without any icon in nav menu).

The "loadError" state could set a state if we want it to.  Same for the other states, so let's go down the list.

"loading", "refreshing", "locked": I think the page status can be silent about these intermediate states.

"loadError", "updateError": These could set the page status so that the user is guided back to the Updates page to review the error messages.

"applying": Should be silent, I'd say.

"restart": Silent as well.

What do you think?

**Comment by mvollmer** [comment]
> Ack, I now [see](https://209.132.184.41:8493/logs/pull-12620-20190910-130303-47b93daf-cockpit-project-cockpit--fedora-30/log.html#30-2) what you meant with the console.error(). So please revert this back to `console.warn()`, as you had before.

Done.

File: pkg/packagekit/updates.jsx (hunk 5 above)
Comment (martinpitt) [reply pkg/packagekit/updates.jsx]: Right, makes sense. It's much less confusing now in the current PR, thanks!

**Comment by mvollmer** [comment]
"needswork" for the page status in error cases.

**Comment by mvollmer** [comment]
We don't need to wait for #12600.  Once we have figured out the design for 12600, we can incorporate OS updates.

**Comment by martinpitt** [comment]
@mvollmer: This now collides with your other PR for insights integration, so this needs a rebase.

**Comment by mvollmer** [comment]
> so this needs a rebase.

Done.

**Comment by martinpitt** [comment]
Force-pushed to fix the conflict. That was related to github showing this weird "rhel-8-0-distropkg" diff.

**Comment by martinpitt** [comment]
@mvollmer: Hmm, I apparently [botched](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-12620-20190921-110101-1e56160a-cockpit-project-cockpit--debian-testing/log.html#236-2) the conflict resolution, can you please have a look?

**Comment by mvollmer** [comment]
@martinpitt: I did the rebase/conflict resolution again, and it's seems to be only about the logic in checl-active-pages for determining the number of preloaded pages.  I gave it my best guess, let's see.

**Comment by mvollmer** [comment]
The "systemd: Use standard colors for notification icons" commit now seems to not be belong here anymore.  I'll split it out.

**Comment by martinpitt** [comment]
The [fedora-30 timeout](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-12620-20190925-064141-ae2ca4ea-cockpit-project-cockpit--fedora-30/log.html) looks worrying, retrying.

**Comment by martinpitt** [comment]
fedora-30 timed out again, there's something broken here..

**Comment by mvollmer** [comment]
> fedora-30 timed out again, there's something broken here..

Looks like the kdump test.

**Comment by mvollmer** [comment]
> > fedora-30 timed out again, there's something broken here..
> 
> Looks like the kdump test.

No, I was confused by the kernel boot output.  That is normal for this test.

**Comment by mvollmer** [comment]
> fedora-30 timed out again, there's something broken here..

The `check-multi-machine-key` test just hangs at some point.  The usual one minute timeout should trigger, but it doesn't.

This is the end of the output:
```
-> wait: ph_is_present("#host-edit-apply:not([disabled])")
> log: PR #host-edit-apply:not([disabled]) 1
<- {'type': 'undefined'}
-> wait: ph_is_visible("#host-edit-apply:not([disabled])")
<- {'type': 'undefined'}
-> ph_mouse("#host-edit-apply:not([disabled])","click",0,0,0,false,false,false,false)
<- {'type': 'undefined'}
-> wait: !ph_is_visible("#host-edit-dialog")
<- {'type': 'undefined'}
-> switch to frame None
-> wait: !ph_is_present("iframe.container-frame[name='cockpit1:10.111.113.2/system']")
> log: PR iframe.container-frame[name='cockpit1:10.111.113.2/system'] 1
> log: PackageKit went away from D-Bus
CDP: executionContextDestroyed 19
CDP: executionContextDestroyed 20
CDP: executionContextDestroyed 21
> log: PackageKit went away from D-Bus
> warning: PackageKit went away during transaction /1_cebdccca : {"problem":"disconnected","command":"close"}
```

And then it hangs forever.

**Comment by mvollmer** [comment]
The test hangs in slightly different places, sometimes also on `wait_not_visible("#host-edit-dialog")`.  There are always the `executionContextDestroyed` messages at the end.

**Comment by mvollmer** [comment]
In a successful run on master, the `executionContextDestroyed` messages also happen, but at a different place in the run.

**Comment by mvollmer** [comment]
When Chromium hangs, it is at 100% CPU.  At that point, it is waiting for the promise returned by client.Runtime.evaluate to resolve.  I added a timeout for that, but `client` seems to be dead afterwards since client.Page.captureScreenshot also just hangs.  The code in cdp-driver.js seems to be happy and unaffected.

**Comment by mvollmer** [comment]
Ok, I am ready to call this a bug in Chromium.

I have tried a couple different versions, including google-chrome, and they all show the same behavior.  Most of the time the test just hangs as described, but very occasionally it goes through.

I am worried that we are starting to stretch the browsers in general with the pre-loaded iframes, what with loading and executing so much stuff in parallel.

Going forward, I'll do two things:
- Find a workaround for the specific tests that are affected. 
- Work on throttling the pre-loading in the shell.

**Comment by mvollmer** [comment]
> I have tried fedora-30 and fedora-31 and both behave the same. We got lucky that fedora-31 passed here.

But then again, the test succeeded on a whole lot of other OSes... hmm.

**Comment by mvollmer** [comment]
The timeouts have disappeared, apparently.  I spin the tests a bit more.

**Comment by martinpitt** [comment]
> Find a workaround for the specific tests that are affected.

We have that now, thanks!
 
> Work on throttling the pre-loading in the shell.

I'm a bit nervous about landing this now as this might affect real users with Chromium? Or is this really just affecting CDP, not interactive usage? I. e. do you want to block this on the trottling?

I'm ok with landing this to get some field testing, of course.

**Comment by mvollmer** [comment]
> > Work on throttling the pre-loading in the shell.
> 
> I'm a bit nervous about landing this now as this might affect real users with Chromium? Or is this really just affecting CDP, not interactive usage? I. e. do you want to block this on the trottling?

Let's not block on the throttling.  I think we are hitting a rare bug in CDP only.
